evap/evaluation/tests/test_tools.py

```diff
@@ -15,3 +16,114 @@ def test_caches_published_course(self):
+        contribution1 = mommy.make(Contribution, contributor=contributor1, course=course, questionnaires=[questionnaire])
+        contribution2 = mommy.make(Contribution, contributor=contributor2, course=course, questionnaires=[questionnaire])
+
+        mommy.make(RatingAnswerCounter, question=question_grade, contribution=contribution1, answer=1, count=1)
+        mommy.make(RatingAnswerCounter, question=question_grade, contribution=contribution2, answer=4, count=2)
+        mommy.make(RatingAnswerCounter, question=question_likert, contribution=contribution1, answer=3, count=4)
+        mommy.make(RatingAnswerCounter, question=question_likert, contribution=general_contribution, answer=5, count=3)
+
+        total_likert = 0.3 * 3 + (1 - 0.3) * 5
+        total_grade = 2.5
+        total = 0.6 * total_grade + (1 - 0.6) * total_likert
+
+        average, deviation = calculate_average_grades_and_deviation(course)
+
+        self.assertAlmostEqual(average, total)
+        self.assertAlmostEqual(deviation, 0)
+
+    @override_settings(CONTRIBUTION_PERCENTAGE=0.3, GRADE_PERCENTAGE=0.6)
+    def test_average_deviation(self):
+        contributor1 = mommy.make(UserProfile)
+        contributor2 = mommy.make(UserProfile)
+
+        course = mommy.make(Course)
+        questionnaire = mommy.make(Questionnaire)
+        question_grade = mommy.make(Question, questionnaire=questionnaire, type="G")
+        question_likert = mommy.make(Question, questionnaire=questionnaire, type="L")
+        general_contribution = mommy.make(Contribution, contributor=None, course=course, questionnaires=[questionnaire])
+        contribution1 = mommy.make(Contribution, contributor=contributor1, course=course, questionnaires=[questionnaire])
+        contribution2 = mommy.make(Contribution, contributor=contributor2, course=course, questionnaires=[questionnaire])
+
+        mommy.make(RatingAnswerCounter, question=question_grade, contribution=contribution1, answer=1, count=1)
+        mommy.make(RatingAnswerCounter, question=question_grade, contribution=contribution1, answer=3, count=1)
+        mommy.make(RatingAnswerCounter, question=question_grade, contribution=contribution2, answer=4, count=2)
+        mommy.make(RatingAnswerCounter, question=question_grade, contribution=contribution2, answer=2, count=2)
+        mommy.make(RatingAnswerCounter, question=question_likert, contribution=contribution1, answer=3, count=4)
+        mommy.make(RatingAnswerCounter, question=question_likert, contribution=contribution1, answer=5, count=4)
+        mommy.make(RatingAnswerCounter, question=question_likert, contribution=general_contribution, answer=5, count=3)
+
+        deviation = calculate_average_grades_and_deviation(course)[1]
+
+        total_likert_dev = 0.3 * 1 + 0.7 * 0
+        total_grade_dev = 1
+        total_dev = 0.6 * total_grade_dev + 0.4 * total_likert_dev
+
+        self.assertAlmostEqual(deviation, total_dev)
```

Comment:
The first one calculates the grade average, the second one calculates the grade deviation. It could be possible to merge them, but that would cause the calculations in each function to be more complex since you can't design the answers to give averages of 1 or similar anymore. Thus, it would reduce readabilty imo.